README.md
Velocity - Template Engine
==========================


[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][coveralls-image]][coveralls-url]
[![npm download][download-image]][download-url]

[npm-image]: http://img.shields.io/npm/v/velocityjs.svg?style=flat-square
[npm-url]: http://npmjs.org/package/velocityjs
[download-image]: https://img.shields.io/npm/dm/velocityjs.svg?style=flat-square
[download-url]: https://npmjs.org/package/velocityjs
[travis-image]: https://img.shields.io/travis/shepherdwind/velocity.js/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/shepherdwind/velocity.js
[coveralls-image]: https://coveralls.io/repos/github/das-buro-am-draht/renderman-velocity-engine/badge.svg?branch=renderman
[coveralls-url]: https://coveralls.io/github/das-buro-am-draht/renderman-velocity-engine?branch=renderman

Velocityjs is [velocity](http://velocity.apache.org/) template engine for javascript.

[中文版文档](./README-cn.md)

## Features

- Supports both client and server side use.
- Separation of parsing and rendering templates.
- The basic syntax is fully supported all java version velocity.
- [Vim Syntax](https://github.com/shepherdwind/vim-velocity) for vim.

## Install

via npm:

```bash
$ npm install velocityjs
```

## Browser

Compatible with all modern browsers. You can try [test case](http://git.shepherdwind.com/velocity.js/runner/tests.html) in your browser to test it.

For other lower version browsers, you need have those polyfill function.

1. Array.prototype map, forEach, some, filter, every, indexOf
2. Date.now
3. Object.keys

## Examples

You can find a lot of examples from the tests directory. There is no different between the use of browser and NodeJs.

## Public API

```
{
  // render method
  render(vm: string, context?: Object, macros?: Object): string;

  parse(vm: string, config?: Object, ignorespace?: boolean): Array<Ast>;

  Compile: {
    (asts: Array<Ast>, config?: Object): {
      render(context?: Object, macros?: Object);
    };
  };
}
```

### render

params:

- vm {string} velocity string input
- context {object} render context, data or function for vm
- macros {object} such as `#include('path/xxx')` , you can define you `inlcude` macro function

```js
var Velocity = require('velocityjs');

Velocity.render('string of velocity', context, macros);
```

#### context

`context` is an object or undefined, for vm `$foo.bar`, data look up path will be `context.foo.bar`.
`context` can have method, and call it just on velocity string.

The method of context, will have `eval` method on `this` of inner method body. You can `eval` to rerender velocity string, such as test code [$control.setTemplate](https://github.com/shepherdwind/velocity.js/blob/master/tests/compile.js#L532).


### Compile and parse

`parse` method can parse vm, and return ast tree of velocity.

`Compile` will render asts to result string.

```
var Compile = Velocity.Compile;

var asts = Velocity.parse('string of velocity');
(new Compile(asts)).render(context, macros);
```

#### Compile

params:

- asts {array} array of vm asts tree
- config {object} you can define some option for Compile

##### config

- escape {boolean} default `false`, default not escape variable to html encode, you can set true to open it.
- unescape {object} define the object, which key do not need escape. For example, set unescape equal `{control: true}`, so `$control.html` will not escape.
- env {string} when env equal `development` will throw error when null values are used
- valueMapper {function} this config allow us to redefine the `#set` value, @see https://github.com/shepherdwind/velocity.js/pull/105

#### parse

params:

- vm {string} string to parse
- blocks {object} self define blocks, such as `#cms(1) hello #end`, you can set `{cms: true}`
- ignorespace {boolean} if set true, then ignore the newline trim.

## Syntax

Syntax you can find from [velocity user guide](http://velocity.apache.org/engine/devel/user-guide.html)。

### Directives

Directives supports have `set`, `foreach`, `if|else|elseif`, `macro`, `break`, `stop`.

Some othe directive `evaluate`, `define`, `parse`, do not supported default, but You can realize by context or macros, for example [parse](https://github.com/shepherdwind/velocity.js/blob/master/tests/compile.js#L627)

## Questions

You can find help from those ways:

1. New [issue](https://github.com/shepherdwind/velocity.js/issues/new)
2. Email to eward.song at gmail.com
3. 阿里内部员工，可以通过 hanwen.sah 搜到我的旺旺

## Other

Recommend an other [velocity](https://github.com/fool2fish/velocity).

## License

(The MIT License)


History.md
## 2.1.1 2020-07-24

### Bug Fixes

- **module:** use correct module name ([4e694ef](https://github.com/das-buro-am-draht/renderman-velocity-engine/commit/4e694ef59b894da452c6c0f5ad67b4df61d24ff1))


## 2.1.0 2020-07-13

### Features

- **compile:** add support for custom invalid reference handler ([4f0a3ef](https://github.com/das-buro-am-draht/renderman-velocity-engine/commit/4f0a3ef2615d4bb749d7124d2cddbd49b1727716))

## 2.0.0 2019-09-26

Break change

First, default escape config set to false. For example

```
#set($foo = '<div>') $foo
``` 

At 1.x , result is `&lt;div&gt;`, for 2.x, output result `<div>`, no escapeHtml variable.

- add quotes to escape when escape option set true [125](https://github.com/shepherdwind/velocity.js/pull/125)

## 1.1.4 2019-08-27

- Added support for remove in ArrayList and Map @mdrewt [#124](https://github.com/shepherdwind/velocity.js/pull/124)
- Support format quiet syntax ${ ref }, fix [#120](https://github.com/shepherdwind/velocity.js/pull/120)

## 1.1.3 2018-09-18

- fixes issue#113 support for add method on arrays by @gauravlanjekar [#114](https://github.com/shepherdwind/velocity.js/pull/114/)

## 1.1.2 2018-08-09

- fix Velocity cli error fix [#109](https://github.com/shepherdwind/velocity.js/issues/109)

## 1.1.1 2018-06-14

- fix: Allow own keySet, entrySet, put and size methods on objects by @lightsofapollo [!106](https://github.com/shepherdwind/velocity.js/pull/106)

## 1.1.0 2018-06-12

- feat: Add support for .put, fix issue [#103](https://github.com/shepherdwind/velocity.js/issues/103) by @lightsofapollo
- feat: add config.valueMapper support by @lightsofapollo [#105](https://github.com/shepherdwind/velocity.js/pull/105)

## 1.0.2 2018-06-06

- fix: ignore path where build on broswer, fix issue [#102](https://github.com/shepherdwind/velocity.js/issues/102)

## 1.0.1 2018-03-08

- fix: #set bug in nested #foreach loops [#100](https://github.com/shepherdwind/velocity.js/issues/100)

## 1.0.0 2018-01-13

- feat: support macro bodyContext [!97](https://github.com/shepherdwind/velocity.js/pull/97)

## 0.10.1 2017-11-10

- fix: #set variable key not work [!91](https://github.com/shepherdwind/velocity.js/issues/92)

## 0.10.0 2017-11-08

- feat: support text version of logical operators [!90](https://github.com/shepherdwind/velocity.js/pull/90)

## 0.9.6 2017-04-14

- fix: keep newline after unparse block [!83](https://github.com/shepherdwind/velocity.js/pull/83)

## 0.9.5 2017-04-07

- fix: support foreach.hasnext when iterating objects [!81](https://github.com/shepherdwind/velocity.js/pull/81)

## 0.9.4 2017-01-16

- fix: #set bug with eval string [#79](https://github.com/shepherdwind/velocity.js/issues/79)

## 0.9.3 2017-01-04

- fix: #set false when in forEach statement [#77](https://github.com/shepherdwind/velocity.js/issues/77)

## 0.9.1 2016-11-22

- add typescript DefinitelyTyped index.d.ts

## 0.9.0 2016-09-20

- feat: support to throw errors in case null values are used [!71](https://github.com/shepherdwind/velocity.js/issues/71)
- fix: support {end} [!70](https://github.com/shepherdwind/velocity.js/issues/71)

## 0.8.5 2016-07-16

- fix bug #foreach with nest empty foreach [!70](https://github.com/shepherdwind/velocity.js/pull/70)

## 0.8.4 2016-07-04

- fix bug when render `#foreach(${itemData} in ${defaultData})`, see
[#69](https://github.com/shepherdwind/velocity.js/issues/69#issuecomment-230152986)

## 0.8.3 2016-06-15

- fix comment bug [#67](https://github.com/shepherdwind/velocity.js/pull/67)

## 0.8.2 2016-05-28

- nothing change, just update npm page readme

## 0.8.0 2016-04-20

- set bug fix [#63](https://github.com/shepherdwind/velocity.js/issues/63)

### Break api change

Before set value will set undefined node as plan object

```
#set($a = {}) #set($a.c.d = 1) $a.c.d
```

This will output `1` .

But now, in 0.8.0 version, a.c will not set as {}, so output is `$a.c.d`.

## 0.7.5

- toString rewrite only when toString equal function { [native code]} [57](https://github.com/shepherdwind/velocity.js/pull/57)

## 0.7.4

- toString like org.apache.velocity when render [56](https://github.com/shepherdwind/velocity.js/pull/56)

## 0.7.3

- fix [53](https://github.com/shepherdwind/velocity.js/issues/52)
- runt support [54](https://github.com/shepherdwind/velocity.js/pull/54)

## 0.7.2

- fix: merge [52](https://github.com/shepherdwind/velocity.js/pull/52)

## 0.7.1

- fix: merge [51](https://github.com/shepherdwind/velocity.js/pull/51)

## 0.7.0

- feat: merge [49](https://github.com/shepherdwind/velocity.js/pull/49) and [#50](https://github.com/shepherdwind/velocity.js/pull/50)

## 0.6.2

- feat: merge [47](https://github.com/shepherdwind/velocity.js/pull/47)

## 0.6.1

- feat: merge [46](https://github.com/shepherdwind/velocity.js/pull/46)

## [0.6](https://github.com/shepherdwind/velocity.js/milestones/0.6)

- change: remove Velocity.Parser, change to Velocity.parse [#43](https://github.com/shepherdwind/velocity.js/issues/43)
- feat: add custom blocks support [#44](https://github.com/shepherdwind/velocity.js/issues/44)

## 0.4

### 0.4.11 / 2015-01-24

- feat: self define macro context keep to origin object

### 0.4.10 / 2015-01-08

- fix: allow optional space after colon in map passed as parameter to macro
      ([#38](https://github.com/shepherdwind/velocity.js/pull/38) by @jamescookie)

### 0.4.9 / 2014-12-29

- feature: support friendly error stack #35
- chore: improve coverage

### 0.4.8 / 2014-12-20

- fix issue #32
- Remove useless code: Helper
- merge pull request #34

### 0.4.7 / 2014-12-18

- fix issue #32


index.d.ts
interface Velocity {
  parse(vmString: string, blocks?: { [blockName: string]: boolean }, ignoreSpace?: boolean): Array<VELOCITY_AST>
  render(vmString: string, context?: RenderContext, macros?: Macros, config?: CompileConfig): string

  Compile: Compile
  Helper: Helper
}

interface Compile {
  new (asts: Array<VELOCITY_AST>, config?: CompileConfig): Compile
  render(context?: RenderContext, macros?: Macros, silence?: boolean): string

  cost: number // render cost time
}

interface Helper {
  getRefText(ast: VELOCITY_AST): string
}

type Macros = Object | { [macro: string]: Function }

type RenderContext = Object | {
  [contextKey: string]: Function,
  eval(vmString: string, context?: RenderContext): string
}

type AST_TYPE = 'set' | 'elseif' | 'if' | 'else' | 'end' | 'foreach' | 'define' | 'content' | 'comment'
                | 'macro' | 'macro_call' | 'math' | 'references' | 'method' | 'index' | 'property'
                | 'bool' | 'interger' | 'decimal' | 'string' | 'array' | 'map'

interface VELOCITY_AST {
  type: AST_TYPE
  value: any
  id: string
  pos: { first_line: number, first_column: number }
}

interface CompileConfig {
  escape?: boolean // escape variable
  unescape?: { [varName: string]: boolean } // unescape var config
  // @see https://github.com/shepherdwind/velocity.js/pull/105
  valueMapper?: (value: any) => any
  invalidReferenceHandler?: (ast: VELOCITY_AST, options?: any) => any
}

declare module 'velocityjs' {
  var velocityjs: Velocity;
  export = velocityjs
}


src/compile/references.js
module.exports = function(Velocity, utils) {

  'use strict';

  function getSize(obj) {

    if (utils.isArray(obj)) {
      return obj.length;
    } else if (utils.isObject(obj)) {
      return utils.keys(obj).length;
    }

    return undefined;
  }

  /**
   * escapeHTML
   */
  function convert(str) {

    if (typeof str !== 'string') return str;

    var result = ""
    var escape = false
    var i, c, cstr;

    for (i = 0 ; i < str.length ; i++) {
      c = str.charAt(i);
      if ((' ' <= c && c <= '~') || (c === '\r') || (c === '\n')) {
        if (c === '&') {
          cstr = "&amp;"
          escape = true
        } else if (c === '"') {
          cstr = "&quot;"
          escape = true
        } else if (c === '<') {
          cstr = "&lt;"
          escape = true
        } else if (c === '>') {
          cstr = "&gt;"
          escape = true
        } else {
          cstr = c.toString()
        }
      } else {
        cstr = "&#" + c.charCodeAt().toString() + ";"
      }

      result = result + cstr
    }

    return escape ? result : str
  }

  function getter(base, property) {
    // get(1)
    if (typeof property === 'number') {
      return base[property];
    }

    var letter = property.charCodeAt(0);
    var isUpper = letter < 91;
    var ret = base[property];

    if (ret !== undefined) {
      return ret;
    }

    if (isUpper) {
      // Address => address
      property = String.fromCharCode(letter).toLowerCase() + property.slice(1);
    }

    if (!isUpper) {
      // address => Address
      property = String.fromCharCode(letter).toUpperCase() + property.slice(1);
    }

    return base[property];
  }

  utils.mixin(Velocity.prototype, {
    /**
     * get variable value 
     * @param {object} ast ast data
     * @param {bool} isVal for example `$foo`, isVal value should be true, other condition,
     * `#set($foo = $bar)`, the $bar value get, isVal set to false
     */
    getReferences: function(ast, isVal) {

      if (ast.prue) {
        var define = this.defines[ast.id];
        if (utils.isArray(define)) {
          return this._render(define);
        }
        if (ast.id in this.config.unescape) ast.prue = false;
      }
      var escape = this.config.escape;

      var isSilent = this.silence || ast.leader === "$!";
      var isfn     = ast.args !== undefined;
      var context  = this.context;
      var ret      = context[ast.id];
      var local    = this.getLocal(ast);

      var text = Velocity.Helper.getRefText(ast);

      if (text in context) {
        return (ast.prue && escape) ? convert(context[text]) : context[text];
      }


      if (ret !== undefined && isfn) {
        ret = this.getPropMethod(ast, context, ast);
      }

      if (local.isLocaled) ret = local['value'];

      if (ast.path) {

        utils.some(ast.path, function(property, i, len) {

          if (ret === undefined) {
            this._throw(ast, property);
          }

          // 第三个参数，返回后面的参数ast
          ret = this.getAttributes(property, ret, ast);

        }, this);
      }

      if (ret === undefined) {
        ret = this.config.invalidReferenceHandler(ast, {isVal: isVal, isSilent: isSilent});
      }

      ret = (ast.prue && escape) ? convert(ret) : ret;

      return ret;
    },

    /**
     * 获取局部变量，在macro和foreach循环中使用
     */
    getLocal: function(ast) {

      var id = ast.id;
      var local = this.local;
      var ret = false;

      var isLocaled = utils.some(this.conditions, function(contextId) {
        var _local = local[contextId];
        if (id in _local) {
          ret = _local[id];
          return true;
        }

        return false;
      }, this);

      return {
        value: ret,
        isLocaled: isLocaled
      };
    },
    /**
     * $foo.bar 属性求值，最后面两个参数在用户传递的函数中用到
     * @param {object} property 属性描述，一个对象，主要包括id，type等定义
     * @param {object} baseRef 当前执行链结果，比如$a.b.c，第一次baseRef是$a,
     * 第二次是$a.b返回值
     * @private
     */
    getAttributes: function(property, baseRef, ast) {
      // fix #54
      if (baseRef === null || baseRef === undefined) {
        return undefined;
      }

      /**
       * type对应着velocity.yy中的attribute，三种类型: method, index, property
       */
      var type = property.type;
      var ret;
      var id = property.id;
      if (type === 'method') {
        ret = this.getPropMethod(property, baseRef, ast);
      } else if (type === 'property') {
        ret = baseRef[id];
      } else {
        ret = this.getPropIndex(property, baseRef);
      }
      return ret;
    },

    /**
     * $foo.bar[1] index求值
     * @private
     */
    getPropIndex: function(property, baseRef) {
      var ast = property.id;
      var key;
      if (ast.type === 'references') {
        key = this.getReferences(ast);
      } else if (ast.type === 'integer') {
        key = ast.value;
      } else {
        key = ast.value;
      }

      return baseRef[key];
    },

    /**
     * $foo.bar()求值
     */
    getPropMethod: function(property, baseRef, ast) {

      var id = property.id;
      var ret = '';

      // getter 处理
      if (id.indexOf('get') === 0 && !(id in baseRef)) {
        if (id.length === 3) {
          // get('address')
          ret = getter(baseRef, this.getLiteral(property.args[0]));
        } else {
          // getAddress()
          ret = getter(baseRef, id.slice(3));
        }

        return ret;

      // setter 处理
      } else if (id.indexOf('set') === 0 && !baseRef[id]) {

        baseRef[id.slice(3)] = this.getLiteral(property.args[0]);
        // $page.setName(123)
        baseRef.toString = function() { return ''; };
        return baseRef;

      } else if (id.indexOf('is') === 0 && !(id in baseRef)) {

        return getter(baseRef, id.slice(2));
      } else if (id === 'keySet' && !baseRef[id]) {

        return utils.keys(baseRef);

      } else if (id === 'entrySet' && !baseRef[id]) {

        ret = [];
        utils.forEach(baseRef, function(value, key) {
          ret.push({key: key, value: value});
        });

        return ret;

      } else if (id === 'size' && !baseRef[id]) {

        return getSize(baseRef);
      } else if (id === 'put' && !baseRef[id]) {
        return baseRef[this.getLiteral(property.args[0])] = this.getLiteral(property.args[1]);
      } else if (id === 'add' && !baseRef[id] && typeof baseRef.push === 'function') {
        return baseRef.push(this.getLiteral(property.args[0]));
      } else if (id === 'remove') {

        if (utils.isArray(baseRef)) {
          
          if (typeof index === 'number') {
            var index = this.getLiteral(property.args[0]);
          } else {
            var index = baseRef.indexOf(this.getLiteral(property.args[0]));
          }

          ret = baseRef[index];
          baseRef.splice(index, 1);
          return ret;

        } else if (utils.isObject(baseRef)) {

          ret = baseRef[this.getLiteral(property.args[0])];
          delete baseRef[this.getLiteral(property.args[0])];
          return ret; 

        }
        
        return undefined;

      } else if (id === 'subList' && !baseRef[id]) {
        return baseRef.slice(this.getLiteral(property.args[0]), this.getLiteral(property.args[1]));
      } else {

        ret = baseRef[id];
        var args = [];

        utils.forEach(property.args, function(exp) {
          args.push(this.getLiteral(exp));
        }, this);

        if (ret && ret.call) {

          var that = this;

          if(typeof baseRef === 'object' && baseRef){
            baseRef.eval = function() {
              return that.eval.apply(that, arguments);
            };
          }

          try {
            ret = ret.apply(baseRef, args);
          } catch (e) {
            var pos = ast.pos;
            var text = Velocity.Helper.getRefText(ast);
            var err = ' on ' + text + ' at L/N ' +
              pos.first_line + ':' + pos.first_column;
            e.name = '';
            e.message += err;
            throw new Error(e);
          }

        } else {
          this._throw(ast, property, 'TypeError');
          ret = undefined;
        }
      }

      return ret;
    },

    _throw: function(ast, property, errorName) {
      if (this.config.env !== 'development') {
        return;
      }

      var text = Velocity.Helper.getRefText(ast);
      var pos = ast.pos;
      var propertyName = property.type === 'index' ? property.id.value : property.id;
      var errorMsg = 'get property ' + propertyName + ' of undefined';
      if (errorName === 'TypeError') {
        errorMsg = propertyName + ' is not method';
      }

      errorMsg += '\n  at L/N ' + text + ' ' + pos.first_line + ':' + pos.first_column;
      var e = new Error(errorMsg);
      e.name = errorName || 'ReferenceError';
      throw e;
    }
  })

}
